lib/common/node.go

```diff
@@ -133,29 +142,40 @@ func (v *Validator) RemoveValidators(validators ...*Validator) error {
 	return nil
 }
 
+// validator has circular reference
+// To marshall/unmarshall the validator, json structure is different
 func (v *Validator) MarshalJSON() ([]byte, error) {
-	var neighbors = make(map[string]struct{})
-	for _, neighbor := range v.validators {
-		neighbors[neighbor.Address()] = struct{}{}
+	var neighbors []string
+	for address, _ := range v.validators {
+		neighbors = append(neighbors, address)
 	}
-	return json.Marshal(map[string]interface{}{
-		"address":    v.Address(),
-		"alias":      v.Alias(),
-		"endpoint":   v.Endpoint().String(),
-		"validators": neighbors,
+
+	return json.Marshal(ValidatorFromJSON{
+		v.Alias(),
+		v.Address(),
+		v.Endpoint().String(),
+		neighbors,
 	})
 }
 
-func (v *Validator) UnmarshalJSON(b []byte) error {
+func (v *Validator) UnmarshalJSON(b []byte) (err error) {
```

Comment:
I figure it's better because `err` variable already declared in return statement(`err error`) when you want to use return variable err
```
func (v *Validator) UnmarshalJSON(b []byte) (err error) {
	var va ValidatorFromJSON
+	if err = json.Unmarshal(b, &va); err != nil {
+		return
	}

	v.alias = va.Alias
	v.address = va.Address
	v.endpoint, err = NewEndpointFromString(va.Endpoint)
	if err != nil {
+		return
	}
	if v.validators == nil {
		v.validators = make(map[string]*Validator)
	}
	for _, neighbor := range va.Validators {
		v.validators[neighbor] = &Validator{}
	}

+	return
}
```